rhombus/scribblings/ref-pair.scrbl

```diff
@@ -103,3 +127,399 @@ its ``rest'' is a list.
+)
+
+}
+
+@doc(
+  ~nonterminal:
+    list_bind: def bind ~defn
+    elem_bind: def bind ~defn
+  bind.macro 'PairList.cons($elem_bind, $list_bind)'
+){
+
+ Matches a non-empty pair list where @rhombus(elem_bind) matches the
+ first element of the pair list and @rhombus(list_bind) matches the
+ rest of the pair list. Static information associated by @rhombus(PairList) is
+ propagated to @rhombus(list_bind).
+
+@examples(
+  def Pair.cons(x, y) = PairList[1, 2, 3]
+  x
+  y
+)
+
+}
+
+
+@doc(
+  def PairList.empty :: PairList = PairList[]
+  bind.macro 'PairList.empty'
+){
+
+  A name and pattern for the empty pair list.
+
+}
+
+
+@doc(
+  fun PairList.first(lst :: NonemptyList)
+){
+
+ Returns the first element of @rhombus(lst).
+ Accessing the first element takes @math{O(1)} time.
+
+@examples(
+  PairList.first(PairList["a", "b", "c"])
+)
+
+}
+
+@doc(
+  fun PairList.rest(lst :: NonemptyList) :: PairList
+){
+
+ Returns a @tech{pair list} like @rhombus(lst), but without its first element.
+ For a given pair list, @rhombus(List.rest) always produces the same value
+ in the sense of @rhombus(===), and accessing that values takes @math{O(1)} time.
+
+@examples(
+  PairList.rest(PairList["a", "b", "c"])
+)
+
+}
+
+
+@doc(
+  ~nonterminal:
+    pair_list_expr: block expr
+  repet.macro 'PairList.repet($list_expr)'
+){
+
+ Creates a repetition from a @tech{pair list}. This is a shorthand for using
+ @rhombus(..., ~bind) with a @rhombus(PairList, ~bind) binding.
+
+@examples(
+  def lst = PairList[1, 2, 3]
+  block:
+    let PairList[x, ...] = lst
+    PairList[x+1, ...]
+  [PairList.repet(lst) + 1, ...]
+)
+
+}
+
+
+@doc(
+  fun PairList.length(lst :: PairList) :: NonnegInt
+){
+
+ Returns the number of items in @rhombus(lst).
+ Computing this length takes @math{O(N)} time.
+
+@examples(
+  PairList.length(PairList[1, 4, 8])
+  PairList.length(PairList[])
+  PairList[1, 4, 8].length
+)
+
+}
+
+
+@doc(
+  fun PairList.reverse(lst :: PairList) :: PairList
+){
+
+ Returns a @tech{pair list} with the same items as @rhombus(lst), but in reversed
+ order. Reversing a list takes @math{O(N)} time.
+
+@examples(
+  PairList.reverse(PairList[1, 4, 8])
+  PairList[1, 4, 8].reverse()
+)
+
+}
+
+
+@doc(
+  fun PairList.append(lst :: PairList, ...) :: PairList
+){
+
+ Appends the @rhombus(lst)s in order. See also @rhombus(++).
+ Appending takes @math{O(N)} time.
+
+@examples(
+  PairList.append(PairList[1, 2, 3], PairList[4, 5], PairList[6])
+  PairList[1, 2, 3].append(PairList[4, 5], PairList[6])
+)
+
+}
+
+
+@doc(
+  fun PairList.take_left(lst :: PairList, n :: NonnegInt) :: PairList
+  fun PairList.take_right(lst :: PairList, n :: NonnegInt) :: PairList
+){
+
+ Like @rhombus(List.take_left) and @rhombus(List.take_right), but for
+ @tech{pair lists}. Creating the new list takes @math{O(log N)} time,
+ where @math{N} is the @rhombus(n) argument for taking from the left or
+ the length of the pair list for taking from the right.
```

Comment:
This description seems wrong; `take_left` creates a new list and takes linear time w.r.t. the number of elements to take (something to note: the implementation doesn’t comply with this!); `take_right` returns a certain tail and takes linear time w.r.t. the length of the list.

Reply:
HF